### PR TITLE
negative range

### PR DESCRIPTION
**Pull request by emre2038**
## Description
bar charts and choropleth can display negative values now

## Related Issue
https://wazimap.atlassian.net/browse/WNCM-571

## How is it tested automatically?
added `__tests__/elements/map/choropleth.test.js`

## How should a reviewer test it locally
* upload a dataset that contains negative values
* confirm that the choropleth, choropleth legend and the charts are displayed correctly

## Screenshots
image
image

## Checklist

- [x]  🚀 is the code ready to be merged and go live?
- [x]  🛠 does it work (build) locally
- [x] 👩‍🎨 does the design match the [Demo](https://wazimap-ng-v1.webflow.io/demo)

### Pull Request

- [x]  📰 good title
- [x]  📝good description
- [x]  🔖 issue linked
- [x]  📖 changelog filled out
- [x] commit messages are meaningful

### Code Quality

- [x]  🚧 no commented out code
- [x]  🖨 no unnecessary logging
- [x]  🎱 no magic numbers

### Testing

- [x]  ✅ added (appropriate) unit tests
- [x]  💢 edge cases in tests were considered
- [ ]  ✅ ran tests locally & are passing

### CODE REVIEW

**Review by goyal1092 — CHANGES_REQUESTED** [review]
Chart will not show up if all values are negative.
https://github.com/OpenUpSA/wazimap-ng-ui/blob/staging/src/js/profile/blocks/indicator.js#L16 will return false.

File: src/js/elements/mapchip/legend.js [hunk 1]
```diff
@@ -32,18 +32,17 @@ export class Legend extends Component {
...
 
         this.isLoading = false;
 
         for (let i = 0; i < intervalLabels.length; i++) {
             const interval = intervalLabels[i];
             const item = this._clonedLegendBlock.cloneNode(true);
-            const label = interval;
 
-            if (i >= lightStart) {
+            if (!isColorLight(colors[i])) {
```
Comment (michaelglenister) [inline src/js/elements/mapchip/legend.js:45]: I like that the comment explaining isColorLight is in the utils file, so it doesn't clutter this code but is easy to find if someone looks deeper

**Review by michaelglenister — COMMENTED** [review]
So far so good, units tests cover the changes. Just one small comment about variable type

If you can address that then this is ready to go

File: src/js/map/maps.js [hunk 2]
```diff
@@ -158,14 +158,38 @@ export class MapControl extends Component {
             calculation
         } = this.choropleth.getChoroplethValues(calculator, childData, primaryGroup, selectedSubindicator, allSubindicators);
 
-        this.choropleth.showChoropleth(calculation, values);
+        const bounds = this.choropleth.getBounds(values);
+        const positiveColorRange = this.choropleth.getColorRange(values, true);
+        const negativeColorRange = this.choropleth.getColorRange(values, false);
+        let zeroColor = this.choropleth.options.zero_color;
+
+        const positiveColorScale = this.choropleth.getScale([bounds.lower, bounds.upper], [positiveColorRange[0], positiveColorRange[1]]);
```
[the remark this answers is not in the document]
Comment (emre2038) [reply src/js/map/maps.js]: great idea - it made sense a lot to me. I updated the code

**Review by michaelglenister — APPROVED** [review]

**Comment by michaelglenister** [comment]
Approved the changes but just remember to have CI passing, or a card to fix it in the backlog

File: src/js/configurations/geography_sa.js [hunk 3]
```diff
@@ -155,6 +155,9 @@ export class Config {
 
         const defaultConfig = {
             colors: d3schemeBlues[5],
+            positive_color_range: ["#eff3ff", "#08519c"],
+            negative_color_range: ["#fef0d9", "#b30000"],
```
Comment (jbothma) [inline src/js/configurations/geography_sa.js:159]: you fell into the same trap as me

here the right-hand item is the more intense colour, but you're using item 0 in the array as the most intense one when there are both positive and negative values.

Comment (jbothma) [reply src/js/configurations/geography_sa.js:159]: so I think the way it's used at https://github.com/OpenUpSA/wazimap-ng-ui/pull/665/files#diff-b2dbf83a3b8c7fe11dcab6b7b935a18c36ffbf277c9e6dbff75a0324de0c03dcR140 is fine - we could document that as "negative max", but then make the defaults be the right order: item 0 negative max, item 1 negative minimum
